src/Data/Json/Extended/Signature/Core.purs

```diff
@@ -0,0 +1,153 @@
+module Data.Json.Extended.Signature.Core
+  ( EJsonF(..)
+  ) where
+
+import Prelude
+
+import Data.Eq1 (class Eq1)
+import Data.Ord1 (class Ord1)
+
+import Data.Foldable as F
+import Data.HugeNum as HN
+import Data.Int as Int
+import Data.List as L
+import Data.Map as Map
+import Data.Tuple as T
+
+-- | The signature endofunctor for the EJson theory.
+data EJsonF a
+  = Null
+  | String String
+  | Boolean Boolean
+  | Integer Int
+  | Decimal HN.HugeNum
+  | String String
+  | Timestamp String
+  | Date String
+  | Time String
+  | Interval String
+  | ObjectId String
+  | Array (Array a)
+  | OrderedSet (Array a)
+  | Object (Array (T.Tuple a a))
+
+instance functorEJsonF ∷ Functor EJsonF where
+  map f x =
+    case x of
+      Null → Null
+      String str → String str
+      Boolean b → Boolean b
+      Integer i → Integer i
+      Decimal a → Decimal a
+      Timestamp ts → Timestamp ts
+      Date d → Date d
+      Time t → Time t
+      Interval i → Interval i
+      ObjectId oid → ObjectId oid
+      Array xs → Array $ f <$> xs
+      OrderedSet xs → OrderedSet $ f <$> xs
+      Object xs → Object $ bimapTuple f <$> xs
+    where
+      bimapTuple f (T.Tuple a b) =
+        T.Tuple (f a) (f b)
+
+instance eq1EJsonF ∷ Eq1 EJsonF where
+  eq1 Null Null = true
+  eq1 (Boolean b1) (Boolean b2) = b1 == b2
+  eq1 (Integer i) (Integer j) = i == j
+  eq1 (Decimal a) (Decimal b) = a == b
+  eq1 (Integer i) (Decimal b) = HN.fromNumber (Int.toNumber i) == b
+  eq1 (Decimal a) (Integer j) = a == HN.fromNumber (Int.toNumber j)
+  eq1 (String a) (String b) = a == b
+  eq1 (Timestamp a) (Timestamp b) = a == b
+  eq1 (Date a) (Date b) = a == b
+  eq1 (Time a) (Time b) = a == b
+  eq1 (Interval a) (Interval b) = a == b
+  eq1 (ObjectId a) (ObjectId b) = a == b
+  eq1 (OrderedSet xs) (OrderedSet ys) = xs == ys
+  eq1 (Array xs) (Array ys) = xs == ys
+  eq1 (Object xs) (Object ys) =
+    let
+      xs' = L.toList xs
+      ys' = L.toList ys
+    in
+      isSubobject xs' ys'
+        && isSubobject ys' xs'
+  eq1 _ _ = false
+
+-- | Very badly performing, but we don't have access to Ord here,
+-- | so the performant version is not implementable.
+isSubobject
+  ∷ ∀ a b
+  . (Eq a, Eq b)
+  ⇒ L.List (T.Tuple a b)
+  → L.List (T.Tuple a b)
+  → Boolean
+isSubobject xs ys =
+  F.foldl
+    (\acc x → acc && F.elem x ys)
+    true
+    xs
+
+instance ord1EJsonF ∷ Ord1 EJsonF where
+  compare1 Null Null = EQ
+  compare1 _ Null = GT
+  compare1 Null _ = LT
+
+  compare1 (Boolean b1) (Boolean b2) = compare b1 b2
+  compare1 _ (Boolean _) = GT
+  compare1 (Boolean _) _ = LT
+
+  compare1 (Integer i) (Integer j) = compare i j
+  compare1 (Integer i) (Decimal b) = compare (HN.fromNumber (Int.toNumber i)) b
+  compare1 (Decimal a) (Integer j) = compare a (HN.fromNumber (Int.toNumber j))
```

Comment:
Would you mind extract this `HN.fromNumber <<< Int.toNumber` to separate function? Or maybe something like `liftInt2 :: forall a. (HugeNumber -> HugeNumber -> a) -> (HugeNumber -> Int -> a)` --> `liftInt2 compare a b`?